Enable Podman and Docker Windows quarkus-container-image-docker testing

Karm opened the pull request:
This PR addresses problems I experienced trying to run `quarkus-container-image-docker` driven tests on Windows with [Podman for Windows](https://github.com/containers/podman/blob/main/docs/tutorials/podman-for-windows.md) and with Docker Desktop for Windows.

I used the changes to Quarkus 2.16 branch to [integrate Quarkus PrimeFaces test](https://github.com/quarkiverse/quarkus-primefaces/pull/12) run that uses a builder image to compile a Linux native executable and then builds a runtime image and runs it for tests. It enables people who work with Windows to test the native workflow despite the fact we do not support Windows MSVC toolchain with e.g. AWT dependency.

# Notes

 * `quarkus.docker.executable-name` and `quarkus.native.container-runtime`, the former actually comes from
    the quarkus-container-image-docker and the latter from the core. It is used instead of the hardcoded "docker"

 * `Caused by: java.nio.file.FileSystemException: target\quarkus.log: The process cannot access the file because it is being used by another process`
    is now wrapped in a try block with a logged warning.

 * Logs from the container: The former logic did not work for me on Windows, the log was left in the container
   and there was nothing to pipe to a file. Was it formerly used with mounting the volume? I added `-i` so as the container
   pours the log to stdout and capture that. WDYT @geoand 🤔?

 * Added quarkus-container-image-docker based test to the AWT extension test to try these changes.

# Testing

I used the AWT integration test like this to test Podman and Docker on Linux and Windows:

```
mvnw clean verify -f integration-tests/pom.xml -pl awt -Ddocker -Dnative
  -Dnative.surefire.skip
  -Dquarkus.native.container-build=true
  -Dquarkus.container-image.build=true
  -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3-java17
  -Dquarkus.native.container-runtime=podman
  -Dquarkus.docker.executable-name=podman
```
(with `docker` in docker variants)

Logs for Quarkus main:

| System     | Podman                        | Docker    |
|------------|-------------------------------|-----------|
| CentOS 8   | 4.0.2 :heavy_check_mark: [logs](https://karms.biz/pastebin/linux-awt-podman-q-main.log.txt)| 20.10.21 :heavy_check_mark: [logs](https://karms.biz/pastebin/linux-awt-docker-q-main.log.txt)|
| Windows 10 | 4.1.0 :heavy_check_mark: [logs](https://karms.biz/pastebin/windows-awt-podman-q-main.log.txt)|  20.10.22 :heavy_check_mark: [logs](https://karms.biz/pastebin/windows-awt-docker-q-main.log.txt) |

Karm committed:
Enable Podman for Windows in-container tests
Enables enforcing podman on systems where both podman and docker are available

Karm committed:
AWT in-container test

Karm commented:
I am on the results, gonna refactor it.

geoand commented:
> Logs from the container: The former logic did not work for me on Windows, the log was left in the container
and there was nothing to pipe to a file. Was it formerly used with mounting the volume? I added -i so as the container
pours the log to stdout and capture that. WDYT @geoand thinking?

I am actually very sceptical of this. I can't remember why I had not ended up doing this, but the only way I will accept this change is if you can guarantee me that it's been thoroughly tested

geoand commented:
The rest of the changes seem perfectly reasonable

geoand commented on test-framework/common/src/main/java/io/quarkus/test/common/DefaultDockerContainerLauncher.java:
```diff
@@ -137,31 +131,31 @@ public void start() throws IOException {
+            Files.deleteIfExists(logFile);
+            Files.createDirectories(logFile.getParent());
+        } catch (FileSystemException e) {
+            log.warnf("Log file %s deletion failed, could happen on Windows, we can carry on.", logFile);
```
The message can probably be limited to Windows

Karm replied:
Will amend.

maxandersen replied:
so on non windows we make it hard fail_

geoand commented on test-framework/common/src/main/java/io/quarkus/test/common/DefaultDockerContainerLauncher.java:
```diff
@@ -24,15 +21,17 @@
 
 public class DefaultDockerContainerLauncher implements DockerContainerArtifactLauncher {
 
+    private static final Logger log = Logger.getLogger(DefaultDockerContainerLauncher.class);
```
Are you sure the logger works properly?

Karm replied:
Well...it logs:
```
[INFO] Running io.quarkus.awt.it.ImageDecodersIT
2023-02-28 12:36:39,352 INFO  [io.qua.tes.com.DefaultDockerContainerLauncher] (main) 
Executing "podman run --name quarkus-integration-test-srEHL -i --rm -p 8081:8081 -p 8444:8444 --net=quarkus-integration-test-qvkwH --env QUARKUS_HTTP_PORT=8081 --env QUARKUS_HTTP_SSL_PORT=8444 --env TEST_URL=http://localhost:8081 karm/quarkus-integration-test-awt:999-SNAPSHOT"
```

So I'd say it's O.K.

zakkak commented on core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java:
```diff
@@ -30,6 +31,20 @@ public static ContainerRuntime detectContainerRuntime() {
         // podman version 2.1.1
         String podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
         boolean podmanAvailable = podmanVersionOutput.startsWith("podman version");
+
+        final String executable = ConfigProvider.getConfig()
+                .getOptionalValue("quarkus.native.container-runtime", String.class).orElse(null);
+        if (executable != null) {
+            if (executable.trim().equalsIgnoreCase("docker") && dockerAvailable) {
+                return ContainerRuntime.DOCKER;
+            } else if (executable.trim().equalsIgnoreCase("podman") && podmanAvailable) {
+                return ContainerRuntime.PODMAN;
+            } else {
+                log.warn("quarkus.native.container-runtime config property must be set to either podman or docker " +
+                        "and the executable must be available. Ignoring it.");
+            }
+        }
+
```
This should be revisited after reaching a conclusion for https://github.com/quarkusio/quarkus/pull/31490#discussion_r1121292851

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java:
```diff
@@ -175,7 +175,7 @@ private static class DockerBinaryStrategy implements Strategy {
         private DockerBinaryStrategy(boolean silent) {
             this.silent = silent;
-            this.binary = ConfigProvider.getConfig().getOptionalValue("quarkus.docker.executable-name", String.class)
+            this.binary = ConfigProvider.getConfig().getOptionalValue("quarkus.native.container-runtime", String.class)
```
Right, there is no good reason to have a different strategy. It might be a good time to have something common

Karm replied:
Note: The thing with `docker.executable-name` is that it's an option introduced in an extension, not in the core, so you might get somewhat confusing: 
```
[WARNING] [io.quarkus.config] Unrecognized configuration key "quarkus.docker.executable-name" 
was provided; it will be ignored; verify that the dependency extension for this configuration is set or
that you did not make a typo
```
I think that the whole "determine container runtime" should happen in the core (podman, docker, alias to one another) and both Test Framework and eventual extensions such as container-image-docker should just consume it.

The unlikely event of having to use one command to use builder image and another command to run containers could still be fine tuned by properties.

Not sure if this PR is the place where this whole thing gets refactored. I'll try.
I am less certain about the 2.16 branch, where going wild on refactoring all that matches word docker might be less desirable.

Karm commented:
Ad CI failures:

```
ResourceTest#testHelloEndpoint() java.util.ServiceConfigurationError:
    io.smallrye.config.SmallRyeConfigFactory:
    io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
```
and
```
io.quarkus.vertx.http.testrunner.UnitET#unitStyleTest(UnitET.java:16) 
    UnitET#unitStyleTest() expected: <Hi> but was: <hello>
io.quarkus.vertx.http.testrunner.UnitET#unitStyleTest2(UnitET.java:10) 
    UnitET#unitStyleTest2() expected: <UNIT> but was: <unit>
```
These are hardly related.

<br>

Ad `PostgresOpenTelemetryJdbcInstrumentationIT`, that feels like it could be very well related, i.e. starting DB containers etc.
It passes on my machine though: https://karms.biz/pastebin/opentelemetry-jdbc-instrumentation-runs-fine.txt

```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13.961 s - in 
io.quarkus.it.opentelemetry.PostgresOpenTelemetryJdbcInstrumentationIT
```

Karm commented on core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java:
```diff
@@ -194,7 +194,7 @@ public Result get() {
                 if (ExecUtil.execWithTimeout(new File("."), filter, Duration.ofMillis(DOCKER_CMD_CHECK_TIMEOUT),
-                        "docker", "version", "--format", "'{{.Server.Version}}'")) {
+                        binary, "version", "--format", "'{{.Server.Version}}'")) {
                     LOGGER.debugf("Docker daemon found. Version: %s", filter.getOutput());
```
@geoand  @zakkak  I can feel the same itch about the abundance of word "Docker" in properties, class names etc.
I just don't know what would be a better alternative for any wide renaming. Should I try "ContainerRuntime" ? Like `IsContainerRuntimeWorking.java`,  `ContainerRuntimeContainerArtifactLauncher...` :smiley:       
Not sure.  I am open to suggestions...   

Leaving it there is IMHO also a perfectly valid option. There are people who just `alias docker=podman` and use that anyway to keep all tools and scripts happy.

geoand replied:
Yeah, it's probably not something for this PR

Karm commented:
> > Logs from the container: The former logic did not work for me on Windows, the log was left in the container
> > and there was nothing to pipe to a file. Was it formerly used with mounting the volume? I added -i so as the container
> > pours the log to stdout and capture that. WDYT @geoand thinking?
> 
> I am actually very sceptical of this. I can't remember why I had not ended up doing this, but the only way I will accept this change is if you can guarantee me that it's been thoroughly tested

I am looking into other projects in Quarkiverse besides PrimeFaces, looking for usages of quarkus-container-image-docker.
I will gradually report what other projects are / are not fine with this change.

Karm committed:
Logs proper command when detecting rootless container runtime

e.g.
Before:

```
WARN  [io.qua.run.uti.ContainerRuntimeUtil] (main) Command "docker" exited with error code 1. Rootless container runtime detection might not be reliable.
```

after:

```
WARN  [io.qua.run.uti.ContainerRuntimeUtil] (main) Command "docker info" exited with error code 1. Rootless container runtime detection might not be reliable or the container service is not running at all.
```
plus with `-Dquarkus.log.level=DEBUG` it logs:

```
DEBUG [io.qua.run.uti.ContainerRuntimeUtil] (main) Command "docker info" output: Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Karm commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:
```diff
@@ -37,6 +38,8 @@ public class AppCDSBuildStep {
     public static final String CLASSES_LIST_FILE_NAME = "classes.lst";
     private static final String CONTAINER_IMAGE_BASE_BUILD_DIR = "/tmp/quarkus";
     private static final String CONTAINER_IMAGE_APPCDS_DIR = CONTAINER_IMAGE_BASE_BUILD_DIR + "/appcds";
+    public static final String DOCKER_EXECUTABLE = ConfigProvider.getConfig()
```
@geoand @zakkak So, going to `ConfigProvider` here is what breaks `integraton-tests\maven` :disappointed:, i.e.

```
2023-03-06 17:34:50,615 ERROR [io.qua.test] (Test runner thread) Test ResourceTest#testHelloEndpoint() failed 
: java.lang.RuntimeException: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
        at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:625)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:696)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:73)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:62)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:363)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:310)
        at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:79)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:286)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$4(ClassBasedTestDescriptor.java:278)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$5(ClassBasedTestDescriptor.java:277)
        at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:31)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:105)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:104)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:68)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$2(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:90)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
        at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
        at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
        at io.quarkus.deployment.dev.testing.JunitTestRunner$2.run(JunitTestRunner.java:227)
        at io.quarkus.deployment.dev.testing.ModuleTestRunner$2.run(ModuleTestRunner.java:92)
        at io.quarkus.deployment.dev.testing.TestSupport.runInternal(TestSupport.java:390)
        at io.quarkus.deployment.dev.testing.TestSupport$2.run(TestSupport.java:310)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
        at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
        at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
        at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:100)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
        at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
        at io.quarkus.test.junit.IntegrationTestUtil.<clinit>(IntegrationTestUtil.java:68)
        at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:231)
        at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:592)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:640)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
        ... 35 more
```

![config](https://user-images.githubusercontent.com/691097/223175267-3d63a3b0-70b5-40cd-bf3d-fafe0cc8b4cf.png)

It doesn't break in regular integration tests, e.g. in the newly added AWT in-container test here. 

I've searched for
```
java.util.ServiceConfigurationError: 
io.smallrye.config.SmallRyeConfigFactory: 
io.quarkus.runtime.configuration.QuarkusConfigFactory 
not a subtype
```
as it rang a bell for me. I have seen it before. It looks like people hit it when doing something that affects classloading. 

I am not sure what to do about it ATM. It is clearer why there was a hardcoded value in the first place though :smile:

Karm replied:
> @geoand @zakkak So, going to `ConfigProvider` here is what breaks `integraton-tests\maven` disappointed, i.e.
> 
> ```
> 2023-03-06 17:34:50,615 ERROR [io.qua.test] (Test runner thread) Test ResourceTest#testHelloEndpoint() failed 
> : java.lang.RuntimeException: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
>         at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:625)
>         at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:696)
>         at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
>         at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
>         at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:73)
>         at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
>         at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
>         at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
>         at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
>         at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
>         at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
>         at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:62)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:363)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:310)
>         at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:79)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:286)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$4(ClassBasedTestDescriptor.java:278)
>         at java.base/java.util.Optional.orElseGet(Optional.java:364)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$5(ClassBasedTestDescriptor.java:277)
>         at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:31)
>         at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:105)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:104)
>         at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:68)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$2(NodeTestTask.java:123)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:123)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:90)
>         at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
>         at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
>         at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
>         at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
>         at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
>         at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
>         at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
>         at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
>         at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
>         at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
>         at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
>         at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
>         at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
>         at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
>         at io.quarkus.deployment.dev.testing.JunitTestRunner$2.run(JunitTestRunner.java:227)
>         at io.quarkus.deployment.dev.testing.ModuleTestRunner$2.run(ModuleTestRunner.java:92)
>         at io.quarkus.deployment.dev.testing.TestSupport.runInternal(TestSupport.java:390)
>         at io.quarkus.deployment.dev.testing.TestSupport$2.run(TestSupport.java:310)
>         at java.base/java.lang.Thread.run(Thread.java:833)
> Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
>         at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
>         at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
>         at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
>         at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
>         at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
>         at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:100)
>         at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
>         at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
>         at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
>         at io.quarkus.test.junit.IntegrationTestUtil.<clinit>(IntegrationTestUtil.java:68)
>         at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:231)
>         at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:592)
>         at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:640)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
>         at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
>         at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
>         at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
>         ... 35 more
> ```
> 
> ![config](https://user-images.githubusercontent.com/691097/223175267-3d63a3b0-70b5-40cd-bf3d-fafe0cc8b4cf.png)
> 
> It doesn't break in regular integration tests, e.g. in the newly added AWT in-container test here.
> 
> I've searched for
> 
> ```
> java.util.ServiceConfigurationError: 
> io.smallrye.config.SmallRyeConfigFactory: 
> io.quarkus.runtime.configuration.QuarkusConfigFactory 
> not a subtype
> ```
> 
> as it rang a bell for me. I have seen it before. It looks like people hit it when doing something that affects classloading.
> 
> I am not sure what to do about it ATM. It is clearer why there was a hardcoded value in the first place though smile


I think I have it nailed. One must not try to call it from a wrong classloader. Running tests locally before I torment Q CI again...

Karm commented:
@geoand  Apart from [the issue](https://github.com/quarkusio/quarkus/pull/31490#pullrequestreview-1326765595) with the `integration-tests\maven`, I didn't see anything like a regression when trying out these. A huge disclaimer though: The projects are hardwired to docker in their own code. So I can see that without this patch, it doesn't even start on a Podman-only system. With this patch, it starts, builds images, all is well until the application decides it needs docker command specifically to do something.

```
https://github.com/quarkiverse/quarkus-mockserver.git
Linux Podman: Works partially, it's hardwired to Docker.
Windows Podman: N/A it's hardwired to Docker in Zeebe code.

https://github.com/quarkiverse/quarkus-zeebe.git
Linux Podman:  N/A it's hardwired to Docker in Zeebe code.
Windows Podman: N/A it's hardwired to Docker in Zeebe code.
Linux Docker:   No regression.

https://github.com/quarkiverse/quarkus-cert-manager.git
Linux Docker:   No regression.
Linux Podman:   No regression.
Windows Podman: No regression.

https://github.com/quarkiverse/quarkus-helm.git
Linux Docker:   No regression.
Linux Podman:   No regression.
Windows Podman: No regression.
```

I tested the  [2.16](https://github.com/quarkusio/quarkus/pull/31491) port of this patch since those projects are not ready for Quarkus 3.0 yet.

geoand commented:
@Karm good to know!

Karm committed:
ConfigProvider.getConfig() different classloader issue

Karm commented:
Much better now. Gonna investigate the last fail...

geoand commented:
The failure looks unrelated to this change

Karm committed:
Container runtime detection cache and rootless check

With this change, the detected container runtime is cached per JVM run
so as it is not checked again regardless of the classloader
loading the ContainerRuntimeUtil class.
The rootless attribute of such runtime is lazily fetched once as needed.

Karm commented:
Commit https://github.com/quarkusio/quarkus/pull/31490/commits/908acff4d5ace758de9db0d5ed8e3ae8b7f1cebd
fixes #31662 , i.e.
 - the docker (podman) info command is executed only once per jvm run and if needed at all
 - the docker version is executed only once per jvm run
 - there is no superflous docker -v or other check...

Karm committed:
Fixes quarkus:dev for env without any docker/podman

Karm commented:
I will see about this next re-run, but so far I have seen Gradle timeouts and Quarkus Scheduler arbitrary failure... Not sure how that would be related.

Karm commented:
@geoand @Sanne  I am happy with this PR now and I'd like it to get merged. 

I realize it's changing stuff, yet that is kind of the point, to change the container runtime detection handling for the better, to make it work on Windows with Podman properly, to get rid of superfluous perpetual calls to docker/podman --version and info etc.

I tested on Linux and Windows and I exhausted what I wanted to check. If you have any suggestion on what to try more or which project to test build with this change, I am all ears.

geoand reviewed (approved):
Thanks a lot!

maxandersen commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java:
```diff
@@ -44,16 +40,19 @@ public NativeImageBuildLocalContainerRunner(NativeConfig nativeConfig, Path outp
-                && ContainerRuntimeUtil.detectContainerRuntime() == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
+        final String selinuxBindOption;
+        if (SystemUtils.IS_OS_MAC && containerRuntime == PODMAN) {
             selinuxBindOption = "";
```
do we know the reason why podman on mac needs different arguments here?

Karm replied:
No idea. I don't have access to any mac hw, so I tested this PR only on Windows and Linux. This bit hasn't been changed though.

n1hility replied:
mac qemu doesnt support selinux labels over 9pfs.  Although not sure if this leads to an actual error, if it does it could be changed to ignore

maxandersen commented on core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java:
```diff
@@ -5,49 +5,134 @@
+     * "quarkus_container_runtime.txt")`
+     * as the file is deleted between JVM executions anyway.
+     */
+    static final Path CONTAINER_RUNTIME = Path.of(System.getProperty("java.io.tmpdir"), "quarkus_container_runtime.txt");
```
why is it it need a file and not just a static var?

gsmet replied:
IIUC, the problem is that we are using different CLs along the build so you end up loading things several time.

That being said, I'm also concerned we are using a file.
First, it won't work well in concurrent build environments as you are using a common file for all builds. That is a bad idea. You need to use `Files.createTempFile()`.
Second, my recollection is that using `deleteOnExit()` is not recommended. So you will need to figure out a better way of deleting the file at the end of the build.

Maybe we need to take a step back here? I'm not sure where this is used but could we use a build item to propagate the value? IIRC the build step will only be executed if it is consumed? Not sure it would be handy though.

gsmet replied:
FWIW, I'm not comfortable releasing Alpha6 with this build concurrency issue in so I'm leaning towards reverting until we come up with a better solution.

So if you feel strongly that I shouldn't, please speak up!

FWIW, the version without the file seemed like a good transition version until we figure things out.

geoand replied:
I don't think the build item will work, because IIRC this is needed in places outside the build steps as well.

That said, using Files.createTempFile() does seem more reasonable

Ladicek replied:
Maybe we could just put it into a `static` field of a class that we always load from the system classloader?

FTR, one such field that is always loaded from the system CL is system properties :laughing:

/me hides

gsmet replied:
So... thinking a bit more about it, using `Files.createTempFile()` in a static field won't work as you will still have different files for different CLs.

But @Ladicek what would be so wrong using a system property? It's a bit of a hack, sure, but it does actually sound more appealing than the other solutions. Would there be any obvious drawback?

geoand replied:
Hack is definitely is... But I think it's probable acceptable. 

On the other hand we could create a temp file based on each JVM run that would not be global to the whole system as the current on is

Karm replied:
@gsmet 
> IIUC, the problem is that we are using different CLs along
> the build so you end up loading things several time.

Yes, exactly. The situation had several stages:
- the current state is really suboptimal, many calls to `-v`, `--version`, `info` to get rootless state etc.
- then I got rid of perpetual `docker info` by having the rootlessness checked lazily only when really needed
- then I got rid of the superfluous `-v` calls, getting the docker command calls to even smaller number
- the last step is caching it in a file. The difference between this and using a static attribute is 1 superfluous docker version call

All in all, if I change the usage of a temporary file to a static variable, the PR would still be very beneficial in this aspect.

> First, it won't work well in concurrent build environments
> as you are using a common file for all builds.

Do you mean there would be a race with `StandardOpenOption.TRUNCATE_EXISTING`? If a build gets an empty file, it merely ignores it and defaults to calling docker again...

> You need to use Files.createTempFile()

I don't think so. That is obviously how temp files are used, but in this context I would end up where I started: Needing to remember somewhere the path to that file. I need to know the path. Not having to remember somewhere `/tmp/container_runtime_<some arbitrary generated number I don't know>.txt`. So that is why I did what I did with the file name.

> recollection is that using deleteOnExit() is not recommended

Could you elaborate? The only problem with deleteOnExit is that is is not guaranteed and it is platform specific, so it's a very bad idea to have it if you need to rely on the file *being really gone*. This is not the case. The worst that might happen (IMHO) is that the file stays there in temp, user switches from docker to podman and there would be a warning or failure depending on what user does. Perhaps we could time stamp it it mention it in the log...

> Maybe we need to take a step back here?
> I'm not sure where this is used but could we use a build item
> to propagate the value?

I am all for taking a step back. I spent some nontrivial time on Windows and Linux, considering dev mode, docker-image extension usage, native build using builder image and test suite using runtime containers. Dumping the info into a file was the thing that looked the least complicated from my limited POV. I am happy to try other avenues.

maxandersen replied:
yeah so build concurency was my concern too about the filename.

Though I'm not sure its that big a deal in practice as what will happen is a concurrent one will pick up the one running on the same system. Only time it will problematic is if those two runs are running with different default environment where one would have picked podman over docker and vice versa. 

One way to avoid that issue is to use `ProcessHandle.current().pid()` and append/prefix the filename. Then at least it will be per process rather than global.

About deleteOnExit() - my understanding is that you just shouldn't assume it always gets deleted (i.e. ctrl+c will hard stop and deletes does not happen); but no other mechanism can do that better anyway.

gsmet replied:
@Karm for now I haven't seen any strong opinions against a system property approach? So if it works, maybe we could try that? That would alleviate all the concerns expressed so far regarding the file approach?

Maybe shoe it in with your current code to see if it works before cleaning things up.

gsmet replied:
Something like `quarkus-local-container-runtime` (lowercase, no `quarkus.` prefix) to make sure it doesn't collide with the config properties.

Karm commented:
@gsmet @maxandersen 

 > One way to avoid that issue is to use `ProcessHandle.current().pid()`
 > and append/prefix the filename. Then at least it will be per process rather than global.

I will add it in another PR and test it on windows, it seems like an obvious thing that just didn't occur to me, Thx.

Karm commented:
@gsmet @Ladicek 

> what would be so wrong using a system property?
> It's a bit of a hack, sure, but it does actually sound more
> appealing than the other solutions. Would there be any obvious drawback?

When I approached that the first time, logging the zillion of various `info` and `-v` an `version` calls with my wrapper:
```
$ cat /home/karm/bin/docker
#!/bin/bash
echo "$(date) docker $@" >> ~/.docker_karm.log
/bin/docker $@
```

My intention was to make the setting last *between* different JVM runs. Then I backed away from that, realizing it would mean yet another garbage left by quarkus runs and there is already enough of that (many vertx things, many jansi files, many temp files etc.).
I briefly considered using `.java` directory in user's home. Then I shrieked in horror realizing that the Quarkus run is deleting and writing something in my home directory.
So the current version does the deleteOnExit thing. All in all, I don't thing there is any obvious drawback in using sys property for that same thing. I can test that...  (It would obviously mean we are giving up on the idea of sharing that config state between JVM runs in any way)

maxandersen commented:
> (It would obviously mean we are giving up on the idea of sharing that config state between JVM runs in any way)

sharing between jvm runs does NOT seem like a good thing IMO.

if we have test runs that can benefit from that its better we make that explicitly share some config.

> the last step is caching it in a file. The difference between this and using a static attribute is 1 superfluous docker version call

if that's the only thing and the pid trick doesn't work then doing static attribute seems just as fine. 1 docker version call sholdn't hurt compared to otherwise fun concurrency debugging a yea from now when we all forgotten we accidentally share state between jvm runs ? :)

Karm commented:
@maxandersen 
> sharing between jvm runs does NOT seem like a good thing IMO.
Well, depends on how you put it. sharing it between arbitrary runs, in `/tmp/container-runtime.txt`, bad idea. Sharing it among the various JVM executions as a part of a single `./mvn verify` call, IMHO O.K.  Maybe the file could be in the project root...? `.container-runtime.txt`

>  1 docker version call
That is per test suite module run AFAIK.

> pid trick doesn't work
I think PID trick would work and I will do it and test it (frowning towards my Windows instance, usual suspects).
I still don't get the concurrency thing though. The file could be in a flux state depending on OS and filesystem:
 - empty
 - not there
 - locked
 - containting garbage
And if that happens, Quarkus merely falls back to calling docker version / podman version again....

Sanne commented:
### Step-back idea 1: follow the example set by Testcontainers?

In Testcontainers, if there is no '~/.testcontainers.properties' file it runs some similar checks, then stores it in the ". file" in unix style - same on Windows.

It's not a perfect solution, I initially actually didn't like it as it led to a puzzling problem for me at some point - but this was quickly fixed once I became aware of this and it's clearly documented.

It does have advantages: it result in consistent behaviour, and once people become aware of it it's simple to manage and to override.

The best feature is that once this is set, the users get the benefit for any subsequent run: across tests, JVMs and projects, and consistently, with the user having clear options.

### Step-back idea 2: actually delegate to Testcontainers code?

Alternatively, and perhaps nicer, is to fully leverage the code and configuration that Testcontainers already requires.

There's a `docker.client.strategy` property in that file, which people might need to have set to get a good experience with podman.

I haven't checked about practicalities in this case, but it certainly sounds appealing to me to delegate the problem:
 - we already expect users to use Testcontainers, so they already had the burden of making sure this is working or configured correctly.
 - don't ask our users to repeat such configuration twice?

gsmet commented:
I think we need:
- a fix for the behavior that was merged, and I would lean towards trying the system property idea for now: that would avoid relying on files (if it works, obviously)
- then we can take a step back and think about the big picture. But we should do that in a second time.

Karm commented:
@Sanne I like the .file situation the most, being it system wide or project wide. Not sure how `~/.testcontainers.properties` address the concern about concurrent execution raised by @gsmet and @maxandersen about the current global `quarkus_container_runtime.txt`. I guess such possibility of having different values for different concurrent runs in 

```
docker.client.strategy=org.testcontainers.dockerclient.UnixSocketClientProviderStrategy
testcontainers.reuse.enable=false
```
is just ignored.

Ad strategy using socket: It did not work for me with Podman on Windows. Not sure if it was permissions (I don't run the TS with elevated rights) or what was it.

@gsmet  O.K., I'll switch to sys prop and show statistic like in https://github.com/quarkusio/quarkus/issues/31662#issuecomment-1465958237
Could you also comment on the possibility of having the file in the project root (and not deleted on JVM exit)? Determined by `basedir` that Maven sets. I use it reliably here:  [Commands.java#L127](https://github.com/Karm/mandrel-integration-tests/blob/master/testsuite/src/it/java/org/graalvm/tests/integration/utils/Commands.java#L127).  Not sure if Gradle sets anything similar...

It must feel like I am obsessed with littering files all over the place.

gsmet commented:
@Karm if you make it a file that survives JVM runs, you completely change the semantic of this feature:

- currently, it's an autodetection feature. If you install Podman, it will detect it, if you go back to Docker, it will detect it.
- if you have a file that survives JVM runs, it won't. So it somehow becomes a configuration file.

If we want to change the semantic, we definitely need to discuss the various options.

aloubyansky commented:
We should fix it in 2.16 too.
